### DISCUSSION TITLE
copy an update file - it disconnects everyone after 15mins.

### DISCUSSION THREAD

**Discussion by dragondaniel** [lloesche/valheim-server-docker#90] [Q&A]
This FIle:
valheim-updater: 

 ./steamcmd.sh +login anonymous +force_install_dir /opt/valheim_dl +app_update 896660 $STEAMCMD_ARGS +quit
    rsync -a --itemize-changes --delete --exclude server_exit.drp --exclude steamapps /opt/valheim_dl/ /opt/valheim | tee "$logfile"
    
    I see this is downling the app and then restarting it.
    
    I added the following line underneath it,  so once it does it rsync I want to copy this specific edit file to replace it.
    I edited the file to increase the server sendrate from default 64kb/s to 512 -- reduce lag for everyone.
     
   " cp /config/assembly_valheim.dll /opt/valheim/valheim_server_Data/Managed/assembly_valheim.dll"
    
    
    
    But after 10-15mins everyone is disconnected, but the docker container is still running, (not destroyed and rebuilt.)
    
    Please can you help me in doing this?
    Cause I want this file to be copied as it increase our bandwidth, so hopefully lag less
    
    
    At the moment I decided to put the cp inside the if statement before the server start up.
    And testing for now.
    
    But if you guys can think of another way, would be great.

**Comment by dragondaniel** [comment]
still kicks everyone. not sure what to do

**Reply by Toucan-Sam** [reply]
I also tried using the modified assembly_valheim.dll and also got crashes, so I removed it.
Some kind of mod support is coming soon (tm) to this container. I would wait until that feature is added instead as this is a feature of ValheimPlus.

**Reply by dragondaniel** [reply]
I think i did something.
But not sure if it is good.

I removed:
--itemize-changes from rsync and exclude the path to that file.

Seems to be working for now...

**Reply by lloesche** [reply]
@Toucan-Sam see https://github.com/lloesche/valheim-server-docker/pull/91
The dev branch `lloesche/valheim-server:dev` now has ValheimPlus support. The config file allows you to modify the `dataRate` variable. If you're brave feel free to test it 🙂 
Or wait for it to be merged into main in a day or two.
Be aware that most settings require the mod to run server as well as client side! The dataRate does not though. Just make sure to also set `enforceMod=false` then so clients are not required to run the mod.

**Reply by Toucan-Sam** [reply]
@lloesche Been running it all night with ~8 people on most of the evening with only the `dataRate` changed. No crashing. Seems functional.

Using the **Map syntax** of docker compose was a bit weird because VALHEIM_PLUS is a true/false and [booleans must be enclosed in quotes](https://github.com/compose-spec/compose-spec/blob/master/spec.md#environment) apparently, so today I learned. Normally just use array but map looks nicer. :)

**Reply by dragondaniel** [reply]
thanks for all this.

I will wait till main.

Not played with docker composed enough.

I am just running this with systemctl on my vm.

Cant wait for this to merged and tried and have a read to learn a bit more :).

**Reply by dragondaniel** [reply]
Off topic:

For file : **valheim-server.service**
          -v /etc/valheim:/config:Z \
          -v /opt/valheim:/opt/valheim:Z \
          
          Why is there two volume mounted?
          Also how come its changed in the systemctl file,
          
          Should it not be the below?
          
    -v $HOME/valheim-server/config:/config \
    -v $HOME/valheim-server/data:/opt/valheim \

**Reply by lloesche** [reply]
@Toucan-Sam that's because bash doesn't know about booleans. the true/false value is really a string and just a convention we're using in our scripts.

**Reply by lloesche** [reply]
@dragondaniel I mean that depends on where you want to store the data. The thinking is, if you install valheim-plus container globally as a systemd service unit on your server, then you might want to store the config and data in global locations like `/etc` and `/opt`. However if you just run it with your personal user from your home directory then that's fine as well. Either is completely valid. All the container cares about is that `/config` is a volume mount so you don't loose your world when you restart the container. Even `/opt/valheim` is optional if you don't mind downloading the 1GB dedicated server binaries from Steam every time the container starts.

**Reply by dragondaniel** [reply]
Hi @lloesche ,

I am still currently using the original and tweaked a bit:

 -v /mnt/valheim/valheim-server-config:/config \
 
 So from my understanding currently my mount poins saves all the config files, worlds, backup etc.
 
 The updater script is to install valheim every time i start up docker.
 
 With this new change to two folders. (config, data)
 
 What is the difference?

**Reply by dragondaniel** [reply]
Not 100% understanding this line:
" Even /opt/valheim is optional if you don't mind downloading the 1GB dedicated server binaries from Steam every time the container starts."


Does it mean if I add this line, it will not download valheim everytime
And
If i do not add the line: it will download valheim everytime?

**Reply by lloesche** [reply]
So that depends on how you use docker. If you always start the same container (retaining the containers filesystem) then it won't download on each startup. But if you're starting a fresh container or run with `docker run --rm` then it would download the server from Steam on each startup unless you volume mount `/opt/valheim`. If that mount is used the server data is persisted on the host filesystem.

**Reply by lloesche** [reply]
To make it short, if you're currently running the container with `-v /mnt/valheim/valheim-server-config:/config` then I recommend adding a `-v /mnt/valheim/valheim-server-data:/opt/valheim` to it. It will speed startups up quite a bit.

**Reply by lloesche** [reply]
> The updater script is to install valheim every time i start up docker

Yes it does that on startup as well as every `UPDATE_INTERVAL` seconds (every 15min by default). That is, it checks for updates, if there is no update it does not download anything.

**Comment by lloesche** [comment]
@dragondaniel if you modify any files in `/opt/valheim/server` (former `/opt/valheim`) then on every update check (every 15m) the Steam updater will "fix" the installation and think that files got update/changed hence restart the server. To work around this all you have to do is define an empty `STEAMCMD_ARGS=""` instead of the default `STEAMCMD_ARGS=validate`.

That said ValheimPlus support has landed in `dev` so replacing the dll should no longer be necessary once it's merged into `main`.
I wrote a whole section about using the mod only for this one specific `dataRate` issue: https://github.com/lloesche/valheim-server-docker/tree/dev#server-data-rate

**Reply by dragondaniel** [reply]
Hi @lloesche ,

Thanks for this.

VALHEIM_PLUS=true

I am guessing this is to be added during the docker run, with the -e option?

**Reply by Toucan-Sam** [reply]
> VALHEIM_PLUS=true
> I am guessing this is to be added during the docker run, with the -e option?

That is correct. You'll see a new `valheimplus/` folder in wherever you've mounted `/config` after the server starts and downloads and installs the mod. Stop the server, update the config file (`valheim_plus.cfg`), restart the server.

EDIT: I ran this all last night and didn't have any issues with disconnection and a relatively high `dataRate` (appropriate for my bandwidth).
